### PR TITLE
Add and apply `focusStyle` function

### PR DESCRIPTION
**Pull request by owi92**
This adds and where possible utilizes a function to apply a default, but somewhat customizable focus outline to elements. This function should be used in place of a manual `:focus` declaration to ensure a uniform focus style that is easily maintainable to Tobira.

Important note:
As noted in #724, Safari does not apply a border radius to outlines. This means, that in Safari, the focus outlines will always be rectangular. But I think this is reasonable enough (see my justification for this below).

I replaced the previously used `box-shadow` with an outline for two reasons:
- `box-shadow` in combination with `outline: none` _can_ be problematic with high contrast modes that use [`forced-colors`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/forced-colors), which forces `box-shadow: none` (for instance Windows high contrast mode). This could be fixed by using sth like `outline: 1px solid transparent` but:
- `outline` is more flexible than `box-shadow`, mainly by allowing to set a specific offset. While `box-shadow` has the `inset` keyword to show it inside the focused element, it does not allow for an outside offset without using multiple shadows.

Closes: #654

### CODE REVIEW

File: frontend/src/layout/header/UserBox.tsx [hunk 1]
```diff
@@ -82,7 +82,7 @@ const LoggedOut: React.FC = () => {
                         color: "var(--nav-color-bw-contrast)",
                     },
                     ":hover": { outline: "2px solid var(--grey80)" },
-                    ":focus": { outline: "2px solid var(--accent-color)" },
+                    ...focusStyle({ offset: 1 }),
```
Comment (LukasKalbertodt) [inline frontend/src/layout/header/UserBox.tsx:85]: The `outlineOffset` a few lines up can be removed then, right?

File: frontend/src/layout/header/Logo.tsx [hunk 2]
```diff
@@ -42,7 +43,7 @@ export const Logo: React.FC = () => {
                 borderRadius: 4,
                 outlineOffset: -2,
                 ":hover": { outline: "2px solid var(--grey80)" },
-                ":focus": { outline: "2px solid var(--accent-color)" },
+                ...focusStyle({ inset: true }),
```
Comment (LukasKalbertodt) [inline frontend/src/layout/header/Logo.tsx:46]: So the `outlineOffset` above can be removed, right?

File: frontend/src/ui/Input.tsx [hunk 3]
```diff
@@ -1,18 +1,15 @@
...
 
 const style = (error: boolean) => ({
     borderRadius: 4,
     border: `1px solid ${error ? "var(--danger-color)" : "var(--grey80)"}`,
-    "&:focus": {
-        outline: "none",
-        boxShadow: "0 0 0 1px var(--accent-color)",
-        borderColor: "var(--accent-color)",
-    },
+    ...focusStyle({}),
```
Comment (LukasKalbertodt) [inline frontend/src/ui/Input.tsx:12]: With this, there is still a grey border right inside of the outline. I think it looks a bit off. I think I would keep the changed border color (although that makes the perceived outline thicker...). Or maybe make the border transparent? Mhhh

File: frontend/src/ui/index.tsx [hunk 4]
```diff
@@ -126,12 +128,21 @@ export const Title: React.FC<TitleProps> = ({ title, className }) => (
     <h2 className={className} css={{ margin: "16px 0" }}>{title}</h2>
 );
 
-export const FOCUS_STYLE_INSET = {
+/**
+ * Applies focus outline with a default width of 2.5px to elements.
+ * This should always be used instead of a custom focus property to ensure
+ * consistency throughout the design.
+ * If ({ inset: true }) is declared, the focus will be on the inside
```
Comment (LukasKalbertodt) [inline frontend/src/ui/index.tsx:135]:
```suggestion
 * If `({ inset: true })` is declared, the focus will be on the inside
```

These docs support markdown.

File: frontend/src/ui/index.tsx [hunk 5]
```diff
@@ -126,12 +128,21 @@ export const Title: React.FC<TitleProps> = ({ title, className }) => (
...
+ * of the focused element with a negative offset equal to the outline's width.
+ * It is also possible to declare an additional offset (usually 1) for
+ * elements with a similar color to the outline to make it stand out more.
+ */
+export const focusStyle = ({ inset = false, offset = 0, width = 2.5 }) => ({
```
Comment (LukasKalbertodt) [inline frontend/src/ui/index.tsx:140]: One could improve the arguments type here. Because specifying `inset` and `offset` at the same time makes no sense, the type should ideally reflect that. Maybe this:

```
type FocusStyleArgs = { width: number } & ({ inset: true } | { offset. number });

const focusStyle = ({ width, ...args }: FocusStyleArgs)
```

**Review by LukasKalbertodt — CHANGES_REQUESTED** [review]
Good to see someone cleaning up this mess that we had before!

One general question: why 2.5 as width? As someone with 200% scaling on all my devices, I don't mind that as it will result in 5 clean physical pixels. But generally one would try to stay with whole pixels to have a better chance of hitting the pixel grid on most devices, resulting in a sharper looking image. Or not? Are browsers free to round to the pixel grid anyway?

File: frontend/src/layout/header/UserBox.tsx (hunk 1 above)
Comment (owi92) [reply frontend/src/layout/header/UserBox.tsx:85]: Yes. This was still used for the hover outline, but I'm debating whether the login button should even have that outline, as it appears to be the only text button with it. I believe I initially added it since the neighboring icon buttons and the search field also have that outline on hover.

So I think I should either remove the hover outline from this or add it to the other text buttons. But in any case, it doesn't need to have an offset.

File: frontend/src/layout/header/Logo.tsx (hunk 2 above)
Comment (owi92) [reply frontend/src/layout/header/Logo.tsx:46]: Yeah, well in this case it is actually necessary for the hover outline, but it can definitely be moved inside the `:hover` state (or that state could as above be removed completely)

File: frontend/src/ui/index.tsx (hunk 5 above)
Comment (owi92) [reply frontend/src/ui/index.tsx:140]: I spent too much time on getting this to work the way I wanted. I had different working versions but here is the shortest I ended up with:
```
type FocusStyleArgs = { width?: number } & ({ inset?: true } | { offset?: number });

export const focusStyle = ({ width = 3, ...args }: FocusStyleArgs) => ({
    "&:focus-visible": {
        outline: `${width}px solid var(--accent-color)`,
        outlineOffset: `${"offset" in args ? args.offset : "inset" in args ? -width : 0}px`,
    },
} as const);
```
I'm not completely sure if that's what you had in mind though.
I think the properties should be optional, so the function can also work without passing any. I'm a little annoyed that I can't access the rest properties directly without checking their existence.
One thing to note is that with this, you're still not actually forbidden to pass both `inset` and `offset` props. An `xor` would be quite nice, but that's not featured in typescript afaik and while it would be possible to implement, I'm not sure we really want those extra lines of code. I guess we could also use a package for that instead, but perhaps it really is sufficient to let developers know via the type that they should only use one of these props.

**Comment by owi92** [comment]
> why 2.5 as width?

Nadine mentioned in the corresponding issue that the focus border is a little too thin and I do agree with that. I tried 3px and actually think that would work well, but might be perceived as a little too thick by some.
So I went with the middle ground without really thinking about it too much, as it worked on my (admittedly relatively high ppi) display. On that it appears that at least firefox and chrome can display even smaller fractions of a pixel, and safari can handle half pixels without sacrificing any sharpness of the image (in my perception). 
I couldn't find any super recent and/or reliable information about this general topic, so I think you might have a point and we should stick with whole numbers for now.

TLDR: I increased this to 3px for now though I wouldn't be totally opposed to stick with 2px instead.

File: frontend/src/layout/header/Logo.tsx (hunk 2 above)
Comment (LukasKalbertodt) [reply frontend/src/layout/header/Logo.tsx:45]: Oh right, I didn't consider that the outline is also used on hover

File: frontend/src/layout/header/UserBox.tsx (hunk 1 above)
Comment (LukasKalbertodt) [reply frontend/src/layout/header/UserBox.tsx:84]: I think removing the hover outline is fine

File: frontend/src/ui/index.tsx (hunk 5 above)
Comment (LukasKalbertodt) [reply frontend/src/ui/index.tsx]: > One thing to note is that with this, you're still not actually forbidden to pass both `inset` and `offset` props.

I'm very surprised by this, but I confirmed just now. Weird. In either case, I probably agree with you that the type shenanigans here are not worth it. So I think reverting it to your original version is totally fine. It's certainly simpler to read.

**Comment by LukasKalbertodt** [comment]
Regarding the non-integer width: I did some experiments aaaand it seems like browsers round to the nearest pixel to get sharp lines! Here are screenshots from Firefox and Chrome for a 2.5px border:

border-width

You likely have to download and then open in a viewer/program where you can zoom with nearest neighbor scaling. My desktop scaling was at 200% for all screenshots, I just zoomed the browser to either 100% or 50%. And well: on the 100% scaling, the border is exactly 2px wide, on the 200% scaling it's 5px wide. 

So... feel very free to put it back to 2.5px :D 
Sorry, sometimes I should rather do the research before I spout some comments into your PRs!

**Comment by LukasKalbertodt** [comment]
> the latest commit (inset for menu item hover) is more of a proposal and I'm not super convinced about that myself, but I'm interested to hear your opinion. Basically I think the "cutoff" between the white arrow tip and greyish menuitem looks just a little weird.

Yes it does look a bit weird, but I don't think it's too much of a problem. Your solution is an interesting one, but with a smaller area changing color and with the bright hover grey, I think the hover effect is too subtle. But we can certainly ask others before deciding on this.

File: frontend/src/layout/header/Search.tsx [hunk 6]
```diff
@@ -107,8 +108,12 @@ export const SearchField: React.FC<SearchFieldProps> = ({ variant }) => {
...
                             paddingLeft: 42,
                             paddingRight: 12,
-                            ":hover": { outline: "2px solid var(--grey80)" },
-                            ":focus": { outline: "2px solid var(--accent-color)" },
+                            ":hover": {
+                                borderColor: "var(--grey80)",
+                                outline: "1.5px solid var(--grey80)",
+                            },
+                            ":focus-visible": { borderColor: "var(--accent-color)" },
+                            ...focusStyle({ width: 1.5 }),
```
Comment (LukasKalbertodt) [inline frontend/src/layout/header/Search.tsx:116]: I found another tiiiiiny problem with this :o  

image

Having the outline and the border right next to one another unfortunately leads to some rendering bugs. I am talking about the 4 slightly bright pixels inside the border there. I have found a solution for this that makes the code even way simpler, so I think it's better for all cases:

```suggestion
                            ...focusStyle({ offset: -1 }),
```

This just puts the outline over the border. 

Whether you do the same trick for the hover outline... I don't care. For those two greys I wouldn't be able to notice the rendering bug.

File: frontend/src/layout/header/UserBox.tsx [hunk 7]
```diff
@@ -118,8 +116,12 @@ const LoggedIn: React.FC<LoggedInProps> = ({ user }) => {
...
                 padding: "8px 10px 8px 16px",
                 cursor: "pointer",
-                ":hover": { outline: "2px solid var(--grey80)" },
-                ":focus": { outline: "2px solid var(--accent-color)" },
+                ":hover": {
+                    borderColor: "var(--grey80)",
+                    outline: "1.5px solid var(--grey80)",
+                },
+                ":focus-visible": { borderColor: "var(--accent-color)" },
+                ...focusStyle({ width: 1.5 }),
```
Comment (LukasKalbertodt) [inline frontend/src/layout/header/UserBox.tsx:124]: Same as previous comment... though this renders fine for me. No wrong pixels. Might still be worth fixing.

File: frontend/src/ui/Blocks/Series.tsx [hunk 8]
```diff
@@ -303,12 +303,9 @@ const FloatingBaseMenu = React.forwardRef<FloatingHandle, FloatingBaseMenuProps>
...
                     whiteSpace: "nowrap",
-                    ":hover, :focus": {
-                        backgroundColor: "var(--grey92)",
-                    },
-                    ":focus-visible": {
-                        outline: "2px solid var(--accent-color)",
-                    },
+                    ":hover, :focus": { backgroundColor: "var(--grey92)" },
+                    ":focus-visible": { borderColor: "var(--accent-color)" },
+                    ...focusStyle({ width: 1.5 }),
```
Comment (LukasKalbertodt) [inline frontend/src/ui/Blocks/Series.tsx:308]: Same once more

File: frontend/src/ui/Input.tsx [hunk 9]
```diff
@@ -1,18 +1,16 @@
...
 const style = (error: boolean) => ({
     borderRadius: 4,
     border: `1px solid ${error ? "var(--danger-color)" : "var(--grey80)"}`,
-    "&:focus": {
-        outline: "none",
-        boxShadow: "0 0 0 1px var(--accent-color)",
-        borderColor: "var(--accent-color)",
-    },
+    ":focus-visible": { borderColor: "var(--accent-color)" },
+    ...focusStyle({ width: 1.5 }),
```
Comment (LukasKalbertodt) [inline frontend/src/ui/Input.tsx:13]: Same again

File: frontend/src/ui/index.tsx [hunk 10]
```diff
@@ -46,6 +44,7 @@ export const LinkList: React.FC<LinkListProps> = ({ items, ...rest }) => (
                 "&:first-child > li:first-child > a": {
                     borderRadius: `${SIDE_BOX_BORDER_RADIUS}px ${SIDE_BOX_BORDER_RADIUS}px 0 0`,
                 },
+                "li:only-child > a": { backgroundColor: "blue", borderRadius: 8 },
```
Comment (LukasKalbertodt) [inline frontend/src/ui/index.tsx:47]: I think you meant to remove the `blue` again :P

**Review by LukasKalbertodt — CHANGES_REQUESTED** [review]

File: frontend/src/ui/index.tsx (hunk 10 above)
Comment (owi92) [reply frontend/src/ui/index.tsx:47]: Whoops... yeah I meant to remove that line

File: frontend/src/layout/header/Search.tsx (hunk 6 above)
Comment (owi92) [reply frontend/src/layout/header/Search.tsx:116]: Oh that is clever, nice.

Comment (LukasKalbertodt) [reply frontend/src/layout/header/Search.tsx]: Mh you only changed the one line in those 4 places, but you didn't remove the `":focus-visible": { borderColor: "var(--accent-color)" }` anywhere. I don't think that line is needed, right?

Comment (owi92) [reply frontend/src/layout/header/Search.tsx]: I thought that would be the case, but then decided to actually check myself this time and found that the border does shine through at the edges. It's hardly noticeable, but I figured it's still better to keep the `borderColor`. (Sure, with the grey border we use at the moment it won't be an issue and I had to change the color to make it apparent, but we might change that in the future as well)
Bildschirm­foto 2023-03-08 um 10 32 01

Bildschirm­foto 2023-03-08 um 10 24 13

Comment (LukasKalbertodt) [reply frontend/src/layout/header/Search.tsx]: Ah, bummer. Good catch. Ye then lets keep it as is.

**Review by LukasKalbertodt — APPROVED** [review]
